### DISCUSSION TITLE
How does sound storage work ?

### DISCUSSION THREAD

**Discussion by Tankeboks** [mcguirepr89/BirdNET-Pi#349] [Q&A] [answered]
I assumed only sound clips with detections were kept, but a few weeks (and 3210 detections) in, my pi managed to fill a 64GB sd card,  I had to clear them out to even get stuff updated so did not check if all the files were ones with detections in them but it does seem like a lot ?

so, my questions : 

- are only clips with hits kept ?
- if not, could there maybe be a function to clear the non-hit clips as a cron job or something?

I guess I could always just dump the sound files to a NAS but if I run this birdpi all summer it could still get quite a large collection of audio files..

**Comment by mcguirepr89** [comment, marked as answer]
Hi, @Klogdor -- thanks for installing BirdNET-Pi -- it sounds like your installation's cleanup and disk_check cron jobs are not working properly, as those should allow your disk to fill only with extractions and will remove old recordings as needed to make room for new recordings (if your "Advanced Settings" "Full Disk Behavior" is set to "Purge").

Also, for perspective, my system that has been running since May 24th has about 7,800 detections and has only used about 2.2G, so that is all just to say that 3,210 detections shouldn't even come close to filling a 64GiB SD card.

Would you mind please providing the output from the command below to see if those cron jobs are indeed running as scheduled:
```
journalctl --no-hostname --no-pager -ucron -n30
```

**Reply by Tankeboks** [reply]
Sure thing : 
```
birdpi-1:~/BirdNET-Pi$ journalctl --no-hostname --no-pager -ucron -n30
 
-- Journal begins at Sat 2022-06-04 08:47:20 CEST, ends at Sat 2022-06-04 17:52:20 CEST. --
Jun 04 17:36:01 CRON[9268]: pam_unix(cron:session): session closed for user robin
Jun 04 17:36:01 CRON[9272]: (robin) CMD (/usr/local/bin/cleanup.sh &> /dev/null)
Jun 04 17:36:01 CRON[9269]: pam_unix(cron:session): session closed for user robin
Jun 04 17:39:02 CRON[10524]: pam_unix(cron:session): session opened for user root(uid=0) by (uid=0)
Jun 04 17:39:02 CRON[10524]: pam_unix(cron:session): session closed for user root
Jun 04 17:40:01 CRON[10815]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:40:01 CRON[10816]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:40:01 CRON[10816]: pam_unix(cron:session): session closed for user robin
Jun 04 17:40:01 CRON[10819]: (robin) CMD (/usr/local/bin/disk_check.sh &> /dev/null)
Jun 04 17:40:01 CRON[10815]: pam_unix(cron:session): session closed for user robin
Jun 04 17:42:01 CRON[11370]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:42:01 CRON[11371]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:42:01 CRON[11373]: (robin) CMD (/usr/local/bin/cleanup.sh &> /dev/null)
Jun 04 17:42:01 CRON[11370]: pam_unix(cron:session): session closed for user robin
Jun 04 17:42:01 CRON[11371]: pam_unix(cron:session): session closed for user robin
Jun 04 17:45:01 CRON[12290]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:45:01 CRON[12291]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:45:01 CRON[12293]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:45:01 CRON[12290]: pam_unix(cron:session): session closed for user robin
Jun 04 17:45:01 CRON[12292]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:45:01 CRON[12291]: pam_unix(cron:session): session closed for user robin
Jun 04 17:45:01 CRON[12292]: pam_unix(cron:session): session closed for user robin
Jun 04 17:45:01 CRON[12299]: (robin) CMD (/usr/local/bin/disk_check.sh &> /dev/null)
Jun 04 17:45:01 CRON[12293]: pam_unix(cron:session): session closed for user robin
Jun 04 17:48:01 CRON[13147]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:48:01 CRON[13147]: pam_unix(cron:session): session closed for user robin
Jun 04 17:50:01 CRON[13740]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
Jun 04 17:50:01 CRON[13743]: (robin) CMD (/usr/local/bin/disk_check.sh &> /dev/null)
Jun 04 17:50:01 CRON[13740]: pam_unix(cron:session): session closed for user robin
Jun 04 17:51:01 CRON[13999]: pam_unix(cron:session): session opened for user robin(uid=1000) by (uid=0)
```

**Reply by Tankeboks** [reply]
it looks like the script is running, but I forgot to mention, I use a RTSP stream in stead of a microphone, could this somehow change things ?

**Reply by mcguirepr89** [reply]
Okay, that much is good -- the next test is to see if there are any errors when running those commands manually:

- **`cleanup.sh`**
  This command will manually run the "cleanup.sh" script with debugging enabled. 
  ```
  sed '/source/a set -x' ~/BirdNET-Pi/scripts/cleanup.sh | bash
  ```
  That command _should_ provide output similar to this:
  ```
  + cd /dev/shm/BirdSongs/Processed
  + empties=($(find ${PROCESSED} -size 57c))
  ++ find /dev/shm/BirdSongs/Processed -size 57c
  + for i in "${empties[@]}"
  + rm -f /dev/shm/BirdSongs/Processed/2022-06-04-birdnet-12:02:49.wav.csv
  + rm -f /dev/shm/BirdSongs/Processed/2022-06-04-birdnet-12:02:49.wav
  ++ find /dev/shm/BirdSongs/Processed
  ++ wc -l
  + [[ 45 -ge 100 ]]
  ```

  Please provide your output here. Before pasting, if you wouldn't mind please formatting your comment for code, it makes things a good bit easier to read. Do this by putting three backticks \`\`\` together on one line and then paste your output on the lines below.
\`\`\`
your output will go here
then close the code block with three more backticks together on their own line at the very bottom
\`\`\`

  creating:
  ```
  your output will go here
  then close the code block with three more backticks together on their own line at the very bottom
  ```
- **`disk_check.sh`**
  This script is a bit trickier to test for efficacy since it really only performs its main task when the disk is truly full. A test can be mimicked by changing the default threshold for what the system considers a "Full Disk" to something that can be immediately tested.
   
   For example:
   To test this on my system, I would first see what the current disk space is, and would adjust the `sed` command below accordingly, making the replacement percentage one (1) percentage less than the current disk size. Note that this will remove some files, but not many and it will start with stuff you probably don't want anyways.
   
   Example:
   ```
   birdnetpi1:~/BirdNET-Pi$ df -h | grep root
   /dev/root        15G  7.9G  5.9G  58% /
   ```
   My disk is 58% full, so to test, I will ask the system to perform the `disk_check.sh` task when the disk is greater than or equal to _57%_, removing 1% of the data needed to bring the space back down to 57%.
   So my `sed` replacement for the test would look like this:
   ```
   sed 's/95/57/g' ~/BirdNET-Pi/scripts/disk_check.sh | bash
   ```
   The output for this command will be abundant if it works and will reveal an error if it fails.


Thank you in advance for running these tests -- it is likely the only way we will get to the bottom of how your system was able to fill.

Let me know if you have any questions~~
Patrick

**Reply by mcguirepr89** [reply]
>it looks like the script is running, but I forgot to mention, I use a RTSP stream in stead of a microphone, could this somehow change things ?

Yes, thank you for letting me know. I assume, now, that this is the direct cause for why your disk filled.

Please provide the output for this command (note it may be ridiculously long -- if so, you can truncate it, I really want to just see what its contents look like in general):

```
source /etc/birdnet/birdnet.conf
ls $RECS_DIR
```

**Reply by ehpersonal38** [reply]
Oops, I guess .WAVs just chucked in $RECS_DIR don't get moved properly to Processed?

**Reply by mcguirepr89** [reply]
Let me do some testing real quick and see what I can find -- the code _looks_ like it would be effective for an RTSP stream, but we'll see.

**Reply by mcguirepr89** [reply]
>Oops, I guess .WAVs just chucked in $RECS_DIR don't get moved properly to Processed?

If the extractions get made (sound file and spectrogram, not database entry), then they definitely get moved to Processed, I'm stumped though. :crossed_fingers: Hopeful that we'll learn something from the output of
```
source /etc/birdnet/birdnet.conf
ls $RECS_DIR
```

**Reply by ehpersonal38** [reply]
Okay I tested with a real RTSP stream and it looks like most things work, though I noticed that the .wavs and their .csvs are moved to `/home/pi/BirdSongs-Analyzed` after being analyzed. Is this the expected directory?

```
pi@birdnetpi:~$ ls -lh
total 6.0M
drwxrwxr-x 11 pi pi 4.0K Jun  4 10:24 BirdNET-Pi
drwxrwxr-x  7 pi pi  20K Jun  4 14:24 BirdSongs
drwxr-xr-x  2 pi pi  20K Jun  4 14:24 BirdSongs-Analyzed
...
```
```
pi@birdnetpi:~$ ls BirdSongs-Analyzed/
2022-06-04-birdnet-14:15:47.wav      2022-06-04-birdnet-14:19:57.wav.csv
2022-06-04-birdnet-14:15:47.wav.csv  2022-06-04-birdnet-14:20:13.wav
2022-06-04-birdnet-14:16:02.wav      2022-06-04-birdnet-14:20:13.wav.csv
2022-06-04-birdnet-14:16:02.wav.csv  2022-06-04-birdnet-14:20:29.wav
2022-06-04-birdnet-14:16:18.wav      2022-06-04-birdnet-14:20:29.wav.csv
2022-06-04-birdnet-14:16:18.wav.csv  2022-06-04-birdnet-14:20:44.wav
2022-06-04-birdnet-14:16:33.wav      2022-06-04-birdnet-14:20:44.wav.csv
2022-06-04-birdnet-14:16:33.wav.csv  2022-06-04-birdnet-14:21:00.wav
2022-06-04-birdnet-14:16:49.wav      2022-06-04-birdnet-14:21:00.wav.csv
2022-06-04-birdnet-14:16:49.wav.csv  2022-06-04-birdnet-14:21:16.wav
2022-06-04-birdnet-14:17:04.wav      2022-06-04-birdnet-14:21:16.wav.csv
2022-06-04-birdnet-14:17:04.wav.csv  2022-06-04-birdnet-14:21:32.wav
2022-06-04-birdnet-14:17:19.wav      2022-06-04-birdnet-14:21:32.wav.csv
2022-06-04-birdnet-14:17:19.wav.csv  2022-06-04-birdnet-14:21:48.wav
2022-06-04-birdnet-14:17:35.wav      2022-06-04-birdnet-14:21:48.wav.csv
2022-06-04-birdnet-14:17:35.wav.csv  2022-06-04-birdnet-14:22:04.wav
2022-06-04-birdnet-14:17:50.wav      2022-06-04-birdnet-14:22:04.wav.csv
2022-06-04-birdnet-14:17:50.wav.csv  2022-06-04-birdnet-14:22:20.wav
2022-06-04-birdnet-14:18:06.wav      2022-06-04-birdnet-14:22:20.wav.csv
2022-06-04-birdnet-14:18:06.wav.csv  2022-06-04-birdnet-14:22:36.wav
...
```

**Reply by ehpersonal38** [reply]
Oops, disregard that last one. Just noticed you made some new commits and they work! RTSP streams first get deposited in /BirdSongs/StreamData, and then get moved to /BirdSongs/Processed.

**Reply by Tankeboks** [reply]
this turned out to be a longer thread than I expected but here you go.

```
robin@birdpi-1:~/BirdNET-Pi$ sed '/source/a set -x' ~/BirdNET-Pi/scripts/cleanup.sh | bash
+ cd /home/robin/BirdSongs/Processed
+ empties=($(find ${PROCESSED} -size 57c))
++ find /home/robin/BirdSongs/Processed -size 57c
++ find /home/robin/BirdSongs/Processed
++ wc -l
+ [[ 1 -ge 100 ]]
```
I'll skip the purge test for now since there have been other developments in the thread and give you the listing, I truncated the ls of $RECS_DIR it a lot , it starts at 

```
2022-06-02-birdnet-20:13:59.wav.csv
2022-06-02-birdnet-20:23:11.wav.csv
2022-06-02-birdnet-20:29:02.wav.csv
2022-06-02-birdnet-20:29:19.wav.csv
2022-06-02-birdnet-20:38:45.wav.csv
2022-06-02-birdnet-20:42:39.wav.csv
2022-06-02-birdnet-21:00:21.wav.csv
2022-06-02-birdnet-21:02:18.wav.csv
2022-06-03-birdnet-01:33:20.wav
2022-06-03-birdnet-01:33:20.wav.csv
2022-06-03-birdnet-01:33:36.wav
2022-06-03-birdnet-01:33:36.wav.csv
```
and ends with 
```
2022-06-05-birdnet-10:13:42.wav
2022-06-05-birdnet-10:13:59.wav
2022-06-05-birdnet-10:14:15.wav
2022-06-05-birdnet-10:14:32.wav
2022-06-05-birdnet-10:14:49.wav
Extracted
May-2022
Processed
```

as a side note, disk usage has gone from 12GB last night to 22 now, and the analyzing job seems to have stopped.

**Reply by Tankeboks** [reply]
> as a side note, disk usage has gone from 12GB last night to 22 now, and the analyzing job seems to have stopped.

Looking at this, it seems that most files (12GB of them) are unprocessed ones..

**Reply by mcguirepr89** [reply]
Hi @Klogdor Please update your system. To process the back log of audio, try running these commands:

```
mkdir -p ~/BirdSongs/StreamData
mv ~/BirdSongs/2022* ~/BirdSongs/StreamData
update_birdnet.sh
```

I think you'll see the backlog begin to be processed at that point. The Update should take care of things going forward.

Please let me know how that goes for you and what types of log output you see in "View Log."

Thanks,
Patrick

**Reply by Tankeboks** [reply]
Allright, ran another update and moved the files and it seems to be analyzing them again now, and any files that show up in the BirdSongs folder seem to be moved immediately, so all good maybe, I'll follow up when it finishes a few days of backlog :D 

The log output looks like this :

```
15:20:19---BirdNET-Pi/birdnet/bin/python3 BirdNET-Pi/scripts\analyze.py --i BirdSongs/2022-06-05-birdnet-15:20:01.wav --o BirdSongs/2022-06-05-birdnet-15:20:01.wav.csv --lat 69.6275 --lon 18.3481 --week 21 --overlap 0.0 --sensitivity 1.25 --min_conf 0.7 --birdweather_id "IN_USE"
15:20:19---ANALYZING AUDIO... DONE! Time 3.2 SECONDS
15:20:19---WRITING RESULTS TO BirdSongs/2022-06-05-birdnet-15:19:44.wav.csv ... DONE! WROTE 0 RESULTS.
15:20:19---READING AUDIO DATA... DONE! READ 5 CHUNKS.
15:20:19---Date-time: 2022-06-05 15:20:01
15:20:22---0.0;3.0-('Tetrao urogallus_Storfugl', 0.099504836)
15:20:22---3.0;6.0-('Tetrao urogallus_Storfugl', 0.0701057)
15:20:22---6.0;9.0-('Tetrao urogallus_Storfugl', 0.10272511)
15:20:22---9.0;12.0-('Tetrao urogallus_Storfugl', 0.070372805)
15:20:22---12.0;15.0-('Tetrao urogallus_Storfugl', 0.07773155)
15:20:23---find: ‘BirdSongs/June-2022/05-Sunday’: No such file or directory
15:20:26---Files loaded
15:20:26---RECORDING_LENGTH set to 15
```

**Reply by Tankeboks** [reply]
actually, when using my eyes, it is only working on the recent files, not touching the backlog files in the StreamData folder.

**Reply by mcguirepr89** [reply]
Thanks for using your eyes :)

I was not thinking earlier -- those files you moved are all analyzed (hence the .csv files), so moving them to `StreamData` should have put them in queue for extraction, and then disk cleanup.

What does your current disk-space look like now?

```
df -h /
```

**Reply by Tankeboks** [reply]
up to 28 GB of usage, from 25 when I last posted. Also, I'm pretty sure they were never actually analyzed (or at least logged in the database) as I have 0 hits on yesterday.. not that it matters with the old stuff as long as it starts running as planned.

I did notice this when running update :

```
+ sudo -u robin git -C /home/robin/BirdNET-Pi reset --hard
HEAD is now at e55d423 Spectrogram should now work again
+ set +x
+ sudo -u robin git -C /home/robin/BirdNET-Pi fetch origin main
error: insufficient permission for adding an object to repository database .git/objects
fatal: failed to write object
fatal: unpack-objects failed
+ set +x
+ sudo -u robin git -C /home/robin/BirdNET-Pi switch -C main --track origin/main
Reset branch 'main'
Branch 'main' set up to track remote branch 'main' from 'origin'.
Your branch is up to date with 'origin/main'.
```

**Reply by mcguirepr89** [reply]
Thanks for letting me know the errors you've encountered for those steps. There was a bug that messed up ownership some of the .git/objects and the .git/index. It was fixed a little while back, but since it broke the updater, you'll need to remedy that before you can get the fix.


Try this:
```
sudo -E chown -R $USER:$USER ~/BirdNET-Pi/.git
update_birdnet.sh
```

and please let me know in what state your system appears to be after the update. Also, do let me know if you see any errors related to the update attempt above.

My best,
Patrick

**Reply by Tankeboks** [reply]
Well, that fixed updating, the system started working on the backlog of files but seems to have stopped during the night, a reboot fixed that for the time, so its working at the time.

so far up to 35GB disk usage.

**Reply by mcguirepr89** [reply]
Hello again, @Klogdor -- something is definitely quite wrong with your system regarding disk-space.

I would be very interested in taking a look at your system if you would be at all amenable to that suggestion -- I can arrange a remote desktop session to look at your installation myself to see if I can determine what is going on.

If you are interested in that troubleshooting, please [email me](mailto:mcguirepr89@gmail.com) and we will arrange a time for a remote session to look at your system together.

If not, I will not get my feelings hurt -- I just wanted to offer since it seems I will learn something about BirdNET-Pi through observing your installation and I'm certain I can get it performing as needed during that session. Also, if you are not interested in a remote session, I do hope you will still keep us apprised of how your system is holding up. I suspect it will fill up in the next few days (or sooner) and will need to be addressed again, anyways.

I'm sorry that my diagnosis for what could be going on is not more comprehensive at this time -- I'm hopeful that we will find the issue sooner than later.

Sincerely,
Patrick

**Reply by Tankeboks** [reply]
Of course you can have a look. I'm out of town until tomorrow and havent set up any remote access to the pi yet, so will mail you when back.

as a side note : it seems to not have stopped overnight and disk usage has gone down to 23GB now (still a day behind on processing)

**Reply by Tankeboks** [reply]
Just an update here ; everything seems to have kept running fine this week, no increase in disk usage and no lag in analyzing, so I guess its healed itself :)
